### DISCUSSION TITLE
android build error on building apk

### DISCUSSION THREAD

**Discussion by arungowda09** [axmolengine/axmol#2260] [Q&A] [answered]
I have a cocos2d project I  have done porting to axmol
version axmol 2.2.1
android studio  version  Android Studio Iguana | 2023.2.1 Patch 1

  ld: error: unable to find library -lext_ssl
  ld: error: unable to find library -lext_crypto
  ld: error: unable to find library -lext_uv
  
  can anybody help me

**Comment by rh101** [comment]
Those errors shouldn't be appearing at all. Are you somehow referencing those libraries from your `CMakeLists.txt`?

**Comment by arungowda09** [comment]
[CMakeLists.txt](https://github.com/user-attachments/files/18063586/CMakeLists.txt)
this is the 3rdparty cmake file

and I have external lib sioclient files that is added in 3rdparty folder but its not added in cmake file but in main project Cmake  reference added

[CMakeLists.txt](https://github.com/user-attachments/files/18063636/CMakeLists.txt)
please check above 2 files

**Reply by rh101** [reply]
> [CMakeLists.txt](https://github.com/user-attachments/files/18063586/CMakeLists.txt) this is the 3rdparty cmake file
> 
> and I have external lib sioclient files that is added in 3rdparty folder but its not added in cmake file but in main project Cmake reference added

~~Where did you originally get this file from before you modified it?~~
Never mind, it seems you copied across the Axmol 3rd party `CMakeLists.txt`, but this is completely unnecessary and not the correct way to be adding a library to your project.

**Reply by rh101** [reply]
To add a library to your project, you would simply add something like the following your project `CMakeLists.txt`:

```
target_link_libraries(${APP_NAME} mylib)
get_target_property(mylib_INCLUDE_DIRS mylib INTERFACE_INCLUDE_DIRECTORIES)
target_include_directories(${APP_NAME}
    PRIVATE ${mylib_INCLUDE_DIRS}
)
```

It needs to be put *after* the section where you see `add_executable(${APP_NAME})`, for example:

```
if(NOT ANDROID)
    add_executable(${APP_NAME} ${APP_SOURCES})
else()
    add_library(${APP_NAME} SHARED ${APP_SOURCES})
    # whole archive for jni
    target_link_libraries(${APP_NAME} -Wl,--whole-archive cpp_android_spec -Wl,--no-whole-archive)
    config_android_shared_libs("org.axmol.lib" "${CMAKE_CURRENT_SOURCE_DIR}/proj.android/app/src")
endif()

# >>>> ADD YOUR LIBRARY HERE <<<<
target_link_libraries(${APP_NAME} mylib)
get_target_property(mylib_INCLUDE_DIRS mylib INTERFACE_INCLUDE_DIRECTORIES)
target_include_directories(${APP_NAME}
    PRIVATE ${mylib_INCLUDE_DIRS}
)
```

**Reply by rh101** [reply, marked as answer]
If you're using libwebsockets, which is included in Cocos2d-x, then you'll need to copy it over to your project along with libuv.

For example, make a folder named "thirdparty" in the root of your project, so it would be `{project_dir}/thirdparty`. 

Now copy `websocket`, `uv`, and `cmake` folders from the Cocso2d-x external folder, and put them in the `{project_dir}/thirdparty` that you created in your project.  The `cmake` folder is required for the file named `CocosExternalConfig.cmake`, which is referenced by the websocket and uv cmake files.

To link these libraries to you project, open your `{project_dir}/CMakeListst.txt`, and modify it to look something like this:

```
if(NOT ANDROID)
    add_executable(${APP_NAME} ${APP_SOURCES})
else()
    add_library(${APP_NAME} SHARED ${APP_SOURCES})
    # whole archive for jni
    target_link_libraries(${APP_NAME} -Wl,--whole-archive cpp_android_spec -Wl,--no-whole-archive)
    config_android_shared_libs("org.axmol.lib" "${CMAKE_CURRENT_SOURCE_DIR}/proj.android/app/src")
endif()

# >>>> ADD YOUR LIBRARIES HERE <<<<
add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/uv ${PROJECT_BINARY_DIR}/app/thirdparty/uv)
target_link_libraries(${APP_NAME} ext_uv)

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/websockets ${PROJECT_BINARY_DIR}/app/thirdparty/websockets)
target_link_libraries(${APP_NAME} ext_websockets)
```

To fix the OpenSSL issue, open `websockets/CMakeLists.txt`, and change these lines:
```
set_property(TARGET ${target_name} APPEND PROPERTY INTERFACE_LINK_LIBRARIES "ext_ssl")
set_property(TARGET ${target_name} APPEND PROPERTY INTERFACE_LINK_LIBRARIES "ext_crypto")
set_property(TARGET ${target_name} APPEND PROPERTY INTERFACE_LINK_LIBRARIES "ext_uv")
```
to:
```
set_property(TARGET ${target_name} APPEND PROPERTY INTERFACE_LINK_LIBRARIES "OpenSSL::SSL")
set_property(TARGET ${target_name} APPEND PROPERTY INTERFACE_LINK_LIBRARIES "OpenSSL::Crypto")
set_property(TARGET ${target_name} APPEND PROPERTY INTERFACE_LINK_LIBRARIES "ext_uv")
```

That should be all. Just be aware that the libwebsocket and libuv that come from Cocos2d-x are old, and may not contain the pre-built libraries for the platform you want to target for your application.  If that is the case, then the only way around this is to build them from source.  Information on how to do would be at the libwebsocket and libuv websites.

**Reply by arungowda09** [reply]
u mean r3dparty cmake file
from axmol engine  library only

**Reply by arungowda09** [reply]
@rh101 
Thank you issue resolved

**Comment by rh101** [comment]
Is there any particular reason you do not want to use the websocket implementation in Axmol?

**Reply by arungowda09** [reply]
Since we've recently updated our project from Cocos2d-x  to Axmol, we're still in the process of evaluating Axmol's WebSocket implementation. At this stage, we're choosing to stick with our current old WebSocket library.
